Implemented Matrix Student's t-distribution

ks147 opened the pull request:
<!-- BEGIN RELEASE NOTES -->
* stats
  * Implemented Matrix Student's t-distribution
<!-- END RELEASE NOTES -->

ks147 committed:
Implemented Matrix t-Distribution

ks147 committed:
Included Example/Documentation for Matrix t-Distribution

ks147 committed:
Fixing empty spacing

ks147 committed:
Improving code quality to pass tests

czgdp1807 commented on sympy/stats/matrix_distributions.py:
```diff
@@ -464,3 +464,106 @@ def MatrixNormal(symbol, location_matrix, scale_matrix_1, scale_matrix_2):
+    def dimension(self):
+        return self.location_matrix.shape
+
+    def pdf(self, x):
+        from sympy import eye
+        nu, M, Omega, Sigma = self.nu, self.location_matrix, self.scale_matrix_1, self.scale_matrix_2
+        n, p = M.shape
+        if isinstance(x, list):
+            x = ImmutableMatrix(x)
+        if not isinstance(x, (MatrixBase, MatrixSymbol)):
+            raise ValueError("%s should be an isinstance of Matrix "
+                             "or MatrixSymbol" % str(x))
```
```suggestion
    def pdf(self, x):
        from sympy import eye
        if isinstance(x, list):
            x = ImmutableMatrix(x)
        if not isinstance(x, (MatrixBase, MatrixSymbol)):
            raise ValueError("%s should be an isinstance of Matrix "
                             "or MatrixSymbol" % str(x))
        nu, M, Omega, Sigma = self.nu, self.location_matrix, self.scale_matrix_1, self.scale_matrix_2
        n, p = M.shape
```

ks147 replied:
I've made the required changes in the most recent commit

ks147 committed:
restructured pdf of matrix student's t-distribution

Smit-create commented on sympy/stats/matrix_distributions.py:
```diff
@@ -464,3 +464,106 @@ def MatrixNormal(symbol, location_matrix, scale_matrix_1, scale_matrix_2):
+
+
+
+def MatrixStudentT(symbol, nu,location_matrix, scale_matrix_1, scale_matrix_2):
+    """
```
```suggestion
def MatrixStudentT(symbol, nu, location_matrix, scale_matrix_1, scale_matrix_2):
    """
```

Smit-create commented on sympy/stats/matrix_distributions.py:
```diff
@@ -464,3 +464,106 @@ def MatrixNormal(symbol, location_matrix, scale_matrix_1, scale_matrix_2):
+    The density of the said distribution can be found at [1].
+
+    Parameters
+    ==========
+    nu: Positive Real number
+        degrees of freedom
```
```suggestion
    ==========
 
    nu: Positive Real number
        degrees of freedom
```

Smit-create commented on sympy/stats/matrix_distributions.py:
```diff
@@ -464,3 +464,106 @@ def MatrixNormal(symbol, location_matrix, scale_matrix_1, scale_matrix_2):
+    [1, 0],
+    [0, 1]]))**(-0.5)/gamma(v/2)
+
+    References
+    ==========
+    .. [1] https://en.wikipedia.org/wiki/Matrix_t-distribution
```
```suggestion
    References
    ==========
    
    .. [1] https://en.wikipedia.org/wiki/Matrix_t-distribution
```

ks147 committed:
Added extra spaces as suggested in code review

Co-authored-by: Smit Lunagariya <55887635+Smit-create@users.noreply.github.com>

ks147 committed:
Removed whitespace

ks147 commented:
@Smit-create @czgdp1807 I've made all the required changes mentioned in code reviews

czgdp1807 commented on sympy/stats/matrix_distributions.py:
```diff
@@ -464,3 +464,109 @@ def MatrixNormal(symbol, location_matrix, scale_matrix_1, scale_matrix_2):
+    @staticmethod
+    def check(nu, location_matrix, scale_matrix_1, scale_matrix_2):
+        if not isinstance(scale_matrix_1, MatrixSymbol):
+            _value_check(scale_matrix_1.is_positive_definite, "The shape "
+                                                              "matrix must be positive definite.")
+        if not isinstance(scale_matrix_2, MatrixSymbol):
+            _value_check(scale_matrix_2.is_positive_definite, "The shape "
+                                                              "matrix must be positive definite.")
+        _value_check(scale_matrix_1.is_square, "Scale matrix 1 should be "
+                                               "be square matrix")
+        _value_check(scale_matrix_2.is_square, "Scale matrix 2 should be "
```
```suggestion
            _value_check(scale_matrix_1.is_positive_definite != False, "The shape "
                                                              "matrix must be positive definite.")
        if not isinstance(scale_matrix_2, MatrixSymbol):
            _value_check(scale_matrix_2.is_positive_definite != False, "The shape "
                                                              "matrix must be positive definite.")
        _value_check(scale_matrix_1.is_square != False, "Scale matrix 1 should be "
                                               "be square matrix")
        _value_check(scale_matrix_2.is_square != False, "Scale matrix 2 should be "
```

czgdp1807 commented:
It's written that in https://en.wikipedia.org/wiki/Matrix_t-distribution that there is no analytic expression for CDF. What happens if we call `expectation` and `variance` on the RandomSymbol following the above distribution? Are the results as expected?

ks147 commented:
@czgdp1807 
This is the output you get if you call Expectation or variance
>>> from sympy.stats import density, MatrixStudentT, E, variance
>>> v = 2
>>> M = MatrixStudentT('M', v, [[1, 2]], [[1, 0], [0, 1]], [1])
>>> E(M)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/evilmorty/Documents/sympy/sympy/stats/rv.py", line 770, in expectation
    return Expectation(expr, condition).doit(**kwargs)
  File "/home/evilmorty/Documents/sympy/sympy/stats/symbolic_probability.py", line 275, in doit
    result = pspace(expr).compute_expectation(expr, evaluate=for_rewrite)
TypeError: compute_expectation() got an unexpected keyword argument 'evaluate'
>>> variance(M)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/evilmorty/Documents/sympy/sympy/stats/rv_interface.py", line 73, in variance
    return cmoment(X, 2, condition, **kwargs)
  File "/home/evilmorty/Documents/sympy/sympy/stats/rv_interface.py", line 229, in cmoment
    return CentralMoment(X, n, condition).doit()
  File "/home/evilmorty/Documents/sympy/sympy/stats/symbolic_probability.py", line 669, in doit
    return self.rewrite(Expectation).doit(**hints)
  File "/home/evilmorty/Documents/sympy/sympy/core/basic.py", line 1769, in rewrite
    return self._eval_rewrite(None, rule, **hints)
  File "/home/evilmorty/Documents/sympy/sympy/core/basic.py", line 1688, in _eval_rewrite
    rewritten = getattr(self, rule)(*args, **hints)
  File "/home/evilmorty/Documents/sympy/sympy/stats/symbolic_probability.py", line 673, in _eval_rewrite_as_Expectation
    return Moment(X, n, mu, condition, **kwargs).rewrite(Expectation)
  File "/home/evilmorty/Documents/sympy/sympy/core/basic.py", line 1769, in rewrite
    return self._eval_rewrite(None, rule, **hints)
  File "/home/evilmorty/Documents/sympy/sympy/core/basic.py", line 1688, in _eval_rewrite
    rewritten = getattr(self, rule)(*args, **hints)
  File "/home/evilmorty/Documents/sympy/sympy/stats/symbolic_probability.py", line 615, in _eval_rewrite_as_Expectation
    return Expectation((X - c)**n, condition)
  File "/home/evilmorty/Documents/sympy/sympy/matrices/expressions/matexpr.py", line 28, in __sympifyit_wrapper
    return func(a, b)
  File "/home/evilmorty/Documents/sympy/sympy/core/decorators.py", line 137, in binary_op_wrapper
    return func(self, other)
  File "/home/evilmorty/Documents/sympy/sympy/matrices/expressions/matexpr.py", line 115, in __sub__
    return MatAdd(self, -other, check=True).doit()
  File "/home/evilmorty/Documents/sympy/sympy/matrices/expressions/matexpr.py", line 97, in __neg__
    return MatMul(S.NegativeOne, self).doit()
  File "/home/evilmorty/Documents/sympy/sympy/matrices/expressions/matmul.py", line 172, in doit
    args = [arg.doit(**kwargs) for arg in self.args]
  File "/home/evilmorty/Documents/sympy/sympy/matrices/expressions/matmul.py", line 172, in <listcomp>
    args = [arg.doit(**kwargs) for arg in self.args]
  File "/home/evilmorty/Documents/sympy/sympy/stats/symbolic_probability.py", line 275, in doit
    result = pspace(expr).compute_expectation(expr, evaluate=for_rewrite)
TypeError: compute_expectation() got an unexpected keyword argument 'evaluate'

czgdp1807 commented:
The above errors are common to all matrix distributions.

ks147 committed:
Update sympy/stats/matrix_distributions.py

Co-authored-by: Gagandeep Singh <gdp.1807@gmail.com>

ks147 commented:
> The above errors are common to all matrix distributions.

So then should I change something in my code ? @czgdp1807 . I think these errors occur because you cannot compare two matrices and thus cannot define Expectation for that reason.

czgdp1807 commented on sympy/stats/matrix_distributions.py:
```diff
@@ -464,3 +464,109 @@ def MatrixNormal(symbol, location_matrix, scale_matrix_1, scale_matrix_2):
+                                                   " of shape %s x %s" % (str(p), str(p)))
+        _value_check(scale_matrix_2.shape[0] == n, "Scale matrix 2 should be"
+                                                   " of shape %s x %s" % (str(n), str(n)))
+        _value_check(nu.is_positive, "Degrees of freedom must be positive")
```
```suggestion
        _value_check(nu.is_positive != False, "Degrees of freedom must be positive")
```
Whenever using checks similar to is_something, test it by `!= False`, so that we are sure that if the required property is doubtful (i.e., returns `None`) we let it pass.

ks147 replied:
Ok I'll be careful about this

ks147 committed:
improving checks

Co-authored-by: Gagandeep Singh <gdp.1807@gmail.com>

czgdp1807 commented:
LGTM. Would merge this tomorrow if no objections raised.
